### PR TITLE
[WIP] center loss

### PR DESCRIPTION
**Pull request by autoih**
Address issue [center loss #643 ](https://github.com/tensorflow/addons/issues/643)

### CODE REVIEW

**Comment by veqtor** [comment]
Nice but I think there's an "alpha" variable that is referenced in the center_loss function that needs to come from somewhere

File: tensorflow_addons/losses/center_loss.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        initializer=tf.constant_initializer(0), trainable=False)
+    labels = tf.reshape(labels, [-1])
+    centers_batch = tf.gather(centers, labels)
+    loss = tf.nn.l2_loss(feature - centers_batch)
+    diff = centers_batch - feature
+    unique_label, unique_idx, unique_count = tf.unique_with_counts(labels)
+    appear_times = tf.gather(unique_count, unique_idx)
+    appear_times = tf.reshape(appear_times, [-1, 1])
+    diff = diff / tf.cast((1 + appear_times), tf.float32)
+    diff = alpha * diff
```
Comment (veqtor) [inline tensorflow_addons/losses/center_loss.py]: alpha?

File: tensorflow_addons/losses/center_loss.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    Args:
+      reduction: (Optional) Type of `tf.keras.losses.Reduction` to apply to
+        loss. Default value is `SUM_OVER_BATCH_SIZE`.
+      name: Optional name for the op.
+    """
+
+    def __init__(self,
+                 reduction=tf.keras.losses.Reduction.AUTO,
+                 feat_dim=3,
+                 num_classes=5,
```
Comment (veqtor) [inline tensorflow_addons/losses/center_loss.py]: Strange to have defaults here, actually it seems feat_dim is unused, also num_classes could be inferred from y_true.shape[-1]

File: tensorflow_addons/losses/center_loss.py (hunk 1 above)
Comment (veqtor) [inline tensorflow_addons/losses/center_loss.py]: Alpha needs to come from somewhere

File: tensorflow_addons/losses/center_loss.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    diff = alpha * diff
+    centers_update_op = tf.scatter_sub(centers, labels, diff)
+
+    return loss, centers, centers_update_op
+
+
+@tf.keras.utils.register_keras_serializable(package='Addons')
+class CenterLoss(tf.keras.losses.Loss):
+    """Computes the center loss.
+
```
Comment (veqtor) [inline tensorflow_addons/losses/center_loss.py]: Perhaps clarify that y_pred needs be a hidden layer representation?

File: tensorflow_addons/losses/center_loss.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                 reduction=tf.keras.losses.Reduction.AUTO,
+                 feat_dim=3,
+                 num_classes=5,
+                 name="center_loss"):
+        super(CenterLoss, self).__init__(reduction=reduction, name=name)
+        self.feat_dim = feat_dim
+        self.num_classes = num_classes
+
+    def call(self, y_true, y_pred):
+        return center_loss(feature, labels, self.num_classes)
```
Comment (veqtor) [inline tensorflow_addons/losses/center_loss.py]: Can't num classes be inferred from y_true.shape[-1]?

File: tensorflow_addons/losses/center_loss.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    def get_config(self):
+        config = {
+            "margin": self.margin,
```
Comment (veqtor) [inline tensorflow_addons/losses/center_loss.py]: This isn't an attribute in this Loss function
Also, missing actual attributes preventing deserialization

**Comment by autoih** [comment]
Thanks @veqtor for useful suggestions. I'll modify the parts you recommended, but I may put this with working in progress temporarily. After I think it's probably okay, I will remove the [wip] tag. What do you think?

File: tensorflow_addons/losses/center_loss.py (hunk 3 above)
Comment (autoih) [reply tensorflow_addons/losses/center_loss.py]: Thanks @veqtor. Updated.

**Comment by autoih** [comment]
Thanks @veqtor. I roughly make changes on the main center loss file. Will work on test cases.

**Comment by veqtor** [comment]
Nice! But still an outdated docstring on the function

File: tensorflow_addons/losses/center_loss.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        feature: feature with shape (batch_size, num_classes)
+        alpha: a scalar between 0-1 to control the leanring rate of the centers
+        num_classes: an `int`. The number of possible classes
+    
+    Return：
+        loss: Tensor
+    """
+    len_features = feature.get_shape()[1]
+    num_classes = labels.shape[-1]
+    centers = tf.get_variable('centers', [num_classes, len_features], dtype=tf.float32,
```
Comment (veqtor) [inline tensorflow_addons/losses/center_loss.py]: This can't possibly work on TF 2.0, right?

File: tensorflow_addons/losses/center_loss.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+"""Implements center loss."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+
+@tf.keras.utils.register_keras_serializable(package='Addons')
+@tf.function
```
Comment (veqtor) [inline tensorflow_addons/losses/center_loss.py]: Maybe not tf.function

File: tensorflow_addons/losses/center_loss.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      reduction: (Optional) Type of `tf.keras.losses.Reduction` to apply to
+        loss. Default value is `SUM_OVER_BATCH_SIZE`.
+    """
+
+    def __init__(self,
+                 reduction=tf.keras.losses.Reduction.AUTO,
+                 alpha=0.2,
+                 name="center_loss"):
+        super(CenterLoss, self).__init__(reduction=reduction, name=name)
+        self.alpha = alpha
```
Comment (veqtor) [inline tensorflow_addons/losses/center_loss.py]: Have centers be a variable here maybe?

File: tensorflow_addons/losses/center_loss.py (hunk 6 above)
Comment (autoih) [reply tensorflow_addons/losses/center_loss.py]: Thanks @veqtor, agree, I'll modify this part.

**Comment by gabrieldemarmiesse** [comment]
@googlebot I consent

**Comment by gabrieldemarmiesse** [comment]
@autoih I've merged master into your branch to update it and fixed any formatting/conflicts it might have. If you need to do some more modifications, please do `git pull` beforehand.

**Comment by gabrieldemarmiesse** [comment]
This pull request was left without activity for a while and the build is not passing. I'm closing it. @autoih  feel free to reopen it if there are changes.
